helix-view/src/editor.rs

```diff
@@ -1077,58 +1078,60 @@ impl Editor {
+        // try to find language servers based on the language name
+        let language_servers = doc.language.as_ref().and_then(|language| {
             self.language_servers
-                .get(language, path.as_ref())
+                .get(language, doc.path())
                 .map_err(|e| {
                     log::error!(
-                        "Failed to initialize the LSP for `{}` {{ {} }}",
+                        "Failed to initialize the language servers for `{}` {{ {} }}",
                         language.scope(),
                         e
                     )
                 })
                 .ok()
-                .flatten()
         });
 
-        let doc = self.document_mut(doc_id)?;
+        // if doc doesn't have a URL it's a scratch buffer, ignore it
         let doc_url = doc.url()?;
 
-        if let Some(language_server) = language_server {
-            // only spawn a new lang server if the servers aren't the same
-            if Some(language_server.id()) != doc.language_server().map(|server| server.id()) {
-                if let Some(language_server) = doc.language_server() {
-                    tokio::spawn(language_server.text_document_did_close(doc.identifier()));
+        if let Some(language_servers) = language_servers {
+            // only spawn new lang servers if the servers aren't the same
+            // TODO simplify?
+            let doc_language_servers = doc.language_servers().collect::<Vec<_>>();
+            let spawn_new_servers = language_servers.len() != doc_language_servers.len()
+                || language_servers
+                    .iter()
+                    .zip(doc_language_servers.iter())
+                    .any(|(l, dl)| l.id() != dl.id());
+            if spawn_new_servers {
+                for doc_language_server in doc_language_servers {
+                    tokio::spawn(doc_language_server.text_document_did_close(doc.identifier()));
                 }
 
                 let language_id = doc.language_id().map(ToOwned::to_owned).unwrap_or_default();
 
-                // TODO: this now races with on_init code if the init happens too quickly
-                tokio::spawn(language_server.text_document_did_open(
-                    doc_url,
-                    doc.version(),
-                    doc.text(),
-                    language_id,
-                ));
+                for language_server in &language_servers {
+                    // TODO: this now races with on_init code if the init happens too quickly
+                    tokio::spawn(language_server.text_document_did_open(
+                        doc_url.clone(),
+                        doc.version(),
+                        doc.text(),
+                        language_id.clone(),
+                    ));
+                }
 
-                doc.set_language_server(Some(language_server));
+                doc.set_language_servers(language_servers);
             }
```

Comment:
Good catch!

Fixed, along with the other suggestions regarding the filtering/retaining order in https://github.com/helix-editor/helix/commit/94754cbb31e272f4f1314ac7b600dd7ba4e46701, please check again.

> I am also very uncomfortable with calling get here. get does a lot of IO internally to figure out where to spawn the LSP (and even worse with #5748 will potentially spawn a new LSP if the filesystem has changed with the last call to get)

I'm not sure if it makes sense to change the behavior in this PR? I tried to stay closer to master.

Reply:
It's fine, I sort of thought about this but I actually relized that this indeed shouldnt be adressed here since the same thjng happens on master. For that reaons I had already edited my comment to remove this, I guess you still had an old version of my comment in your

Reply:
I need to look more closely at that code again but it looks it be going in the right direction